### ISSUE TITLE
enable AppInit for non-wow64 on win8+

### ISSUE THREAD

**Issue by derekbruening** [DynamoRIO/dynamorio#1522]
_From [bruen...@google.com](https://code.google.com/u/109494838902877177630/) on August 21, 2014 11:05:01_

issue #1035 covers enabling AppInit for WOW64 on win8+.  We blanket-disabled
AppInit for win8+ but we should be able to enable for non-WOW64.  This
issue covers that.

_Original issue: http://code.google.com/p/dynamorio/issues/detail?id=1522_

**Comment by derekbruening** [comment]
r2788 solved the static import check problem.  There are other problems, however: a hang at exit is the most prominent.

**Comment by derekbruening** [comment]
If I remove drpreinject unloading itself => now it works for WOW64.
DR in there already: also works.

This unload-itself hack is done for all apps, even those unconfigured for, and I do see hangs at exit running user32-using apps w/ the self-unload.

We should just switch to returning false from DllMain if on win8+, and live w/ drpreinject remaining in the address space if loaded from shimeng.

Confirmed that returning false is getting us unloaded for no-inject into calc wow64:

Unloaded modules:
74da0000 74dba000   DRPREI~1.DLL

x64 seems to have further issues: but wow64 AppInit now works regularly and in mixed-mode on win8.1 (after the #1035 fixes).

**Comment by TomSie** [comment]
Hey Derek,

as I reported in #4283, one of the bugs in AppInit on x64 win8+ is a missaligned stack. This is my proposed fix:

```
--- core/arch/pre_inject_asm.asm
+++ core/arch/pre_inject_asm.asm
@@ -114,6 +114,8 @@ do_not_unload:
          */
         mov      eax, 0
 #ifdef X64
+        /* repair the stack and return */
+        pop      rcx
         ret
 #else
         ret      HEX(0c)
```

Applying this fix will fix the situation that all processes crash (even those that are not meant to be run under dynamorio). It will, however, not fix crashes for processes that are meant to be run under dynamorio.

**Comment by derekbruening** [comment]
OK, could you send a pull request?

**Comment by TomSie** [comment]
I tried, but got `Permission to DynamoRIO/dynamorio.git denied to TomSie`

**Comment by derekbruening** [comment]
> I tried, but got `Permission to DynamoRIO/dynamorio.git denied to TomSie`

Are you following https://github.com/DynamoRIO/dynamorio/wiki/Code-Reviews#submitting-a-change-as-a-non-project-member ?

**Comment by TomSie** [comment]
Hey,
no I didn't know about that process. I cloned the repo and created the pull request as documented:

https://github.com/TomSie/dynamorio/pull/1

Is that how it's supposed to work?
Regards

**Comment by derekbruening** [comment]
> Hey,
> no I didn't know about that process. I cloned the repo and created the pull request as documented:
> 
> [TomSie/dynamorio#1](https://github.com/TomSie/dynamorio/pull/1)
> 
> Is that how it's supposed to work?

Not quite: the pull request should merge into DynamoRIO/dynamorio rather than your repository.  Are you sure you forked it from DynamoRIO/dynamorio?  A forked repo should offer the origin repo as a default choice for a new pull request to merge into.

**Comment by TomSie** [comment]
Like this? https://github.com/DynamoRIO/dynamorio/pull/4294

**Comment by derekbruening** [comment]
> Like this? #4294

Yes, thank you for the patch.

**Comment by derekbruening** [comment]
As I put in the PR we could also use more AppInit tests including for non-takeover.
